src/transports/smart.c

```diff
@@ -66,6 +66,76 @@ static int git_smart__set_callbacks(
 	return 0;
 }
 
+#define forbid_custom_header(disallowed_name) \
+	if (strncmp(disallowed_name, custom_header, name_len) == 0) \
+		return false
+
+bool is_valid_custom_header(const char *custom_header)
+{
+	const char *c;
+	int name_len;
+
+	if (custom_header == NULL)
+		return true;
+
+	// Disallow \r and \n
+	c = strchr(custom_header, '\r');
+	if (c != NULL)
+		return false;
+	c = strchr(custom_header, '\n');
+	if (c != NULL)
+		return false;
+
+	// Require a header name followed by :
+	c = strchr(custom_header, ':');
+	if (c == NULL)
+		return false;
+	name_len = c - custom_header;
+	if (name_len < 1)
+		return false;
+
+	// Disallow headers that we set
+	forbid_custom_header("User-Agent");
+	forbid_custom_header("Host");
+	forbid_custom_header("Accept");
+	forbid_custom_header("Content-Type");
+	forbid_custom_header("Transfer-Encoding");
+	forbid_custom_header("Content-Length");
+
+	return true;
+}
+
+const char *find_invalid_custom_header(const git_strarray *custom_headers)
+{
+	size_t i;
+
+	if (custom_headers == NULL || custom_headers->count == 0)
+		return NULL;
+
+	for (i = 0; i < custom_headers->count; i++)
+		if (!is_valid_custom_header(custom_headers->strings[i]))
+			return custom_headers->strings[i];
+
+	return NULL;
+}
+
+static int git_smart__set_custom_headers(
+	git_transport *transport,
+	const git_strarray *custom_headers)
+{
+	transport_smart *t = (transport_smart *)transport;
+	const char *invalid_header = find_invalid_custom_header(custom_headers);
+
+	if (invalid_header != NULL) {
+		giterr_set(GITERR_INVALID, "Illegal HTTP header '%s'", invalid_header);
```

Comment:
This message would suggest that `Host` is an ilegal header, since we consider it invalid. It would be much clearer if we could distinguish between a header field we don't allow to override and one which the user passed in malformed.

The error message should not start with a capital letter.